### DISCUSSION TITLE
tutorials/lift-off-part2/08-server-codegen

### DISCUSSION THREAD

**Discussion by giscus** [apollographql-education/content-feedback#59] [course feedback]
# tutorials/lift-off-part2/08-server-codegen

https://www.apollographql.com/tutorials/lift-off-part2/08-server-codegen

**Comment by khobaib222** [comment]
1. In previous lesson we used the ```documents``` keyword in the codegen file to indicate which type of files should be read for generating the types. Why isn't this keyword in used in this lesson for server side config?
2. In this lesson we are using ```@graphql-codegen/typescript``` plugin as base to generate typescript types, why this plugin was not included in the first lesson when we generated types for client side?
3. What happens when lets say our ```Track``` is being resolved from different parent resolvers which send different type of data to resolve the ```author``` field. So lets say one parent sends ```authorId``` and another sends ```authorName``` to resolve the author field using different apis, for this type of case should we define two different models or should be define a single model with optional properties ```authorName``` and ```authorId```? And also do we need to put multiple if else conditions in our Track resolver if they are being resolved from different parents sending different data to resolve the track?

**Reply by lizhennessy** [reply]
Hi there @khobaib222 ! Great questions.

1. The `documents` keyword: We used this property on the _client_ side because we wanted to tell the code generator to scan for any documents that contained GraphQL operations. On the frontend, these were any of the queries that we wrote to ask our GraphQL server for data - such as `GetTracks`! This enables the Code Generator to first identify all of the operations we plan on using and create the relevant types we can use to anticipate the kind of data we'll get in response. We don't need to use this `documents` property on the _server_ side, because we don't actually write any GraphQL operations _here_. Instead, we simply tell the Code Generator where our schema file lives with the `schema` property, and it can generate all the types we need to use in our server files. Here's the [documentation on that property](https://the-guild.dev/graphql/codegen/docs/config-reference/documents-field) if you want to read up on it!
2. The `@graphql-codegen/typescript` package - this is a great question! We have a number of different packages we can use with the Code Generator based on what we are trying to do. On the client side, we have some client-specific configurations that let us attune our code generator right to our project and the different dependencies it includes. In the case of the server side, the [`typescript` package](https://the-guild.dev/graphql/codegen/plugins/typescript/typescript) is actually just the base package required for some server plugins - such as the `typescript-resolvers` package we installed. Essentially, we install different packages on the client and the server side based on our specific use case; the ones we installed on the client still generated the relevant TypeScript types that we need to use in our React code! I have linked the documentation for the typescript package above if you're interested.
3. The `Track` model: in this case, you could define your `TrackModel` type as a union type that points to both of the definitions you defined: such as `type TrackModel = TrackWithAuthorId | TrackWithAuthorName`. These types could be defined individually within the same file, and contain all of the relevant `Track` fields, in addition to the one that varies!

I hope that this has helped to answer some of your questions - please let us know if there's anything else we can help with!

**Comment by GeorgeOrzanescu** [comment]
Following along and got hit by this when running  ```npm run generate```. 

Error: Unable to find Codegen config file! 

        Please make sure that you have a configuration file under the current directory!


Codegen file is identical, under src folder. Any ideas ? Thank you !

**Reply by GeorgeOrzanescu** [reply]
Hmm this worked from the official docs ```npx graphql-code-generator --config ./path/to/codegen.ts```.

**Reply by MikMazayeu** [reply]
codegen.ts should be in server folder, not in server/src

**Comment by neogeek13** [comment]
What am I missing here? Looks like the output file is in quotes to me:
______________________
...
  generates: {
    "./src/types.ts": {},
  },
};

export default config;
______________________
Note: The path to the output file is not contained in quotes, and it ends with a colon (:) - this is because we're about to specify some additional properties beneath it!

**Reply by lizhennessy** [reply]
Thanks for calling this typo out @neogeek13 ! You're right - that "not" shouldn't be there! We'll fix it right away.

**Comment by neogeek13** [comment]
"The type of the parent argument being passed to this function has updated to be TrackModel. Our codegen is complete!"

TrackModel?? Not string? When I hover over the parent argument { authorId } I see:
(parameter) authorId: string

**Reply by lizhennessy** [reply]
Hey there @neogeek13 - you're absolutely right, seeing the `authorId` property typed as a `string` is accurate. And I agree, it's a little unclear that the parent argument is actually a `TrackModel` type.  If you hover over `{ dataSources }` you'll probably see the same sort of limited information. We can tell that it's an object with a `trackAPI` property, but not much else.

Instead, if you hover over the `author` resolver name itself, you'll see a more comprehensive type signature that gives us a few more hints about what's going in this function.

<img width="1250" alt="image" src="https://github.com/apollographql-education/content-feedback/assets/95302380/162cf17e-fb95-498d-b12c-9e4ffa736252">

The first few bits are there to clarify that this is a `Resolver` type that we're working with, and there's some GraphQL Codegen specifics in the middle there. We can see that the `Resolver` type is a generic, and it's passed three things:

1. `ResolverTypeWrapper<AuthorModel>`
2. `TrackModel`
3. `DataSourceContext`

This tells us that the resolver function for `Track.author` actually has quite a lot going into it to make up a complex type. If we jump into the `types.ts` file we've generated, we'll learn a bit more.

Here's the typing for `ResolverTypeWrapper`:

```
export type ResolverTypeWrapper<T> = Promise<T> | T;
```

So it looks like this type is also a generic that receives an argument `T` - in our example, that's the `AuthorModel` - and we can see this is indeed what our resolver function does; when we ask for a `Track.author`, we expect it to be a Promise that resolves to an Author instance.

To understand how the `Resolver` type is actually using the `TrackModel` and `DataSourceContext` that are being passed into it, we can jump to the `Resolver` definition in our `types.ts` file.  I'm only going to paste in the first bit of the type definition because it gets a bit hairy, and I think here's where we can see the gist of it --

```export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}>```

So we can see from this signature that the `Resolver` type actually receives a few different things - a type that describes the result, i.e. the kind of data it'll return, a type to describe the parent argument, one for the context argument, and finally one for the arguments! Our `Track.author` function doesn't have any arguments, which is why we don't see it in the type definition.

(If you want to explore this a little deeper, you can jump down to the `ResolverFn` type in the `types.ts` file - it shows how those arguments are actually applied to our resolver function as their types.)

Phew! I hope that wasn't too much of an explanation. In this case, it's really the cost AND the benefit of setting the type for our resolvers map at the top of `resolvers.ts` - it accounts for everything that happens in our resolver functions, obscuring away a lot of the little details. But they're still there for us to discover if we dig into the `types.ts` file that is generated from our schema/resolvers/datasources/models. :) Please let us know if you have any other questions!

**Reply by neogeek13** [reply]
Yeah I'd seen the TrackModel on the larger scope hover but I guess I assumed we were referencing the hover over of the specific (parent) argument vs. the overall view. Thanks for the detailed review! :)

**Comment by belloshehu** [comment]
Obviously, this is the hardest and blurry part so far. It is exciting though how via the Codegen configuration, types definition and inferences are resolved. If I may ask, does one have to use the Codegen?

**Reply by Theedon** [reply]
No, you don't but it is extremely helpful and can save you a lot of debugging time in the future

**Comment by Brandon-Perry** [comment]
I ran into an issue where after running codegen, it looks like my types file is missing some types. I have a type called Doc, but when I get to the 'function useGetDocQuery' part at the end of the file, it says that the GetDocQuery and GetDocQueryVariables types are missing. Any idea what might be causing this?

**Reply by mabuyo** [reply]
Hey there @Brandon-Perry - can you clarify what you mean by `Doc` type? Is this a new type you're adding to the schema?

**Reply by Brandon-Perry** [reply]
Yes, Doc is a type that I've defined in my schema. Here's what my files look like:
```ts
// types.ts (generated file)

// truncated
export const GetDocDocument = gql`
    query GetDoc {
  doc {
    id
    text
    created_at
  }
}
    `;

/**
 * __useGetDocQuery__
 *
 * To run a query within a React component, call `useGetDocQuery` and pass it any options that fit your needs.
 * When your component renders, `useGetDocQuery` returns an object from Apollo Client that contains loading, error, and data properties
 * you can use to render your UI.
 *
 * @param baseOptions options that will be passed into the query, supported options are listed on: https://www.apollographql.com/docs/react/api/react-hooks/#options;
 *
 * @example
 * const { data, loading, error } = useGetDocQuery({
 *   variables: {
 *   },
 * });
 */
export function useGetDocQuery(baseOptions?: Apollo.QueryHookOptions<GetDocQuery, GetDocQueryVariables>) {
        const options = {...defaultOptions, ...baseOptions}
        return Apollo.useQuery<GetDocQuery, GetDocQueryVariables>(GetDocDocument, options);
      }
// truncated
```
Here's my schema
```ts
// schema.ts
import { gql } from "@apollo/client";

export const typeDefs = gql`
  type Query {
    hello: String
    doc: Doc
  }

  scalar Date

  type User {
    id: ID
    doc: Doc
  }

  type Doc {
    id: ID
    text: String
    created_at: Date
  }
`
```

my resolvers
```ts
// resolvers.ts
import { Resolvers } from "./types";

export const resolvers: Resolvers = {
  Query: {
    hello: () => "hello",
    doc: () => {
      return {
        id: "",
        text: "Test",
        created_at: new Date().toISOString(),
      };
    },
  },

};
```
my queries
```ts
import { gql } from "@apollo/client";

export const queries = gql`
    query GetDoc {
        doc {id text created_at}
    }
`
```
and finally my codegen
```ts
import { CodegenConfig } from "@graphql-codegen/cli";

const config: CodegenConfig = {
  schema: ["./src/lib/graphql/schema.ts", "./src/lib/graphql/resolvers"],
  generates: {
    "./src/lib/graphql/types.ts": {
      plugins: ["typescript", "typescript-resolvers", "typescript-react-apollo"],
    },
  },
  documents: ["./src/**/*.ts"],
  require: ['ts-node/register'],
  config: {
    scalars: {
      Date: "string",
    },

  }
};

export default config;
```
It will generate my useGetDocQuery and it will work at getting my data. However, type inference doesn't seem to be working.

**Reply by BlenderDude** [reply]
A couple things:

1. It looks like you are using both the client _and_ resolver plugins at once and on the same file. I'd highly recommend splitting this into two generated files.

```ts
import { CodegenConfig } from "@graphql-codegen/cli";

const config: CodegenConfig = {
  schema: ["./src/lib/graphql/schema.ts", "./src/lib/graphql/resolvers"],
  generates: {
    "./src/lib/graphql/resolvers.ts": {
      plugins: ["typescript", "typescript-resolvers"],
    },
    "./src/lib/graphql/client.ts": {
      preset: 'client' // <-- see below
     }
  },
  documents: ["./src/**/*.ts"],
  require: ['ts-node/register'],
  config: {
    scalars: {
      Date: "string",
    },
  }
};

export default config;
```

2. I would highly recommend using the client preset in liu of `typescript-react-apollo` _especially_ in a project where client and server code is mixed. Instead of emitting a `useGetDocQuery()` it emits `GetDoc` as a `TypedDocumentNode` which is a GraphQL document with extra Typescript type information injected into it. You will need to install the `@graphql-codegen/client-preset` package.

If you want to expand on your use-case, I'd be more than happy to dive in further if you encounter more issues. It seems like both client and server code are in the same place, so I suspect this is a NextJS (or similar) application where your client and server live together. I personally recommend splitting these into two separate apps, as GraphQL generally serves as the strict communication layer, and codegen allows you to get generated types on both sides (client and server) without any of their code touching.

**Comment by Tsimopak** [comment]
So if I add new type, I should make a Model for it and add it to codegen? Is there a way to skip the part of letting codegen where to search so for example to use name convention or annotate beside each type with its unique id name.

**Comment by mark-P0** [comment]
In the codegen docs there is a section about Apollo Server and the use of `useIndexSignature` config. However it is not used in this part of the tutorial. Is it safe to ignore that?

Link to the docs: https://the-guild.dev/graphql/codegen/plugins/typescript/typescript-resolvers#integration-with-apollo-server

Thank you for the tutorial

**Reply by lizhennessy** [reply]
Thanks for the question @mark-P0 ! I am looking into this. I can see the reason why we'd want [index signatures in TypeScript](https://www.linkedin.com/pulse/typescript-index-signatures-4-examples-type-safe-dynamic-efimenko-u0ivf), however I didn't encounter any issues _not_ including them in the codegen file. I'll investigate and update the course accordingly if it makes the generated types more efficient and future-proof! Thank you for bringing this to our attention. :)

**Comment by Dyrits** [comment]
How can I generate directly from Prisma ?

```ts
import { City, Cuisine, Manager, Restaurant } from "@prisma/client";

export interface RestaurantModel extends Restaurant {}
export interface CityModel extends City {}
export interface ManagerModel extends Manager {}
export interface CuisineModel extends Cuisine {}
```

Then I mapped the models. Is it a good way?

**Reply by lizhennessy** [reply]
Hi there @Dyrits ! Typically I've tried to decouple my GraphQL schema types from the models used in a Prisma DB. This can be useful when we don't want every aspect of our database model setup to be reflected in our schema. When coupling this with TypeScript, I'll create a `models.ts` file where I include the types that represent data objects extracted from the database. I generally include those models in the codegen process, for example:

```
mappers: {
    Restaurant: "../datasources/models#RestaurantModel"
}
```

This gives our resolver functions an understanding of the types of properties available to us as prescribed by the `RestaurantModel`, for example.

 In tandem with this, I've usually created a class with methods I use to access data from the DB. This can help me with further transformations. Is there a reason you're creating interfaces that extend particular types?

**Reply by Dyrits** [reply]
I created the interface just to not mix up the types names, using the TypeModel nomination for the models coming from Prisma, and Type nomination, for the types coming from GraphQL (auto-generated).

**Comment by samuelpucat** [comment]
Exciting, I was wondering when the typescript part would come in.
Since we are using REST API as one of our data sources, and that is defined by swagger/openapi, can we generate models for example with [openapi-typescript](https://www.npmjs.com/package/openapi-typescript)? Or is it better to define them manually?

**Reply by lizhennessy** [reply]
Hi there @samuelpucat ! I definitely wouldn't say that manual definition is _better_, only that it's a good option to be familiar with! If you already have a breakdown of the types in your REST API, that sounds like a great way to save some time. :)

**Comment by ASJ98** [comment]
Should a model type match what is returned from the REST API and I return exactly that in the resolver, and then add separate field resolvers for any schema field that does not exactly match the field name from the REST API? Or should I modify the REST API response to suit my schema and thus give the model different fields?
